### PR TITLE
cleanup: Simplify usage of MinIOSourceProxyRequest

### PR DESCRIPTION
**Pull request by rluetzner**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description

This replaces a convoluted condition that ultimately evaluated to

"is this HTTP header present in the request or not"

, similar to what happens with MinIOSourceReplicationRequest a few lines further down in `cmd/object-api-options.go`.

We only care that the HTTP header is set, the actual value of it is irrelevant.

## Motivation and Context

Just cleaning up.

## How to test this PR?

-

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/object-api-options.go [hunk 1]
```diff
@@ -66,9 +66,8 @@ func getDefaultOpts(header http.Header, copySource bool, metadata map[string]str
 	if crypto.S3.IsRequested(header) || (metadata != nil && crypto.S3.IsEncrypted(metadata)) {
 		opts.ServerSideEncryption = encrypt.NewSSE()
 	}
-	if v, ok := header[xhttp.MinIOSourceProxyRequest]; ok {
-		opts.ProxyHeaderSet = true
-		opts.ProxyRequest = strings.Join(v, "") == "true"
+	if _, ok := header[xhttp.MinIOSourceProxyRequest]; ok {
+		opts.ProxyRequest = true
```
Comment (harshavardhana) [inline cmd/object-api-options.go]:
```suggestion
		_, opts.ProxyRequest = header[xhttp.MinIOSourceProxyRequest]
```

you wanted to simplify :-)

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/object-api-options.go (hunk 1 above)
Comment (rluetzner) [reply cmd/object-api-options.go]: Good catch. 🙂

**Review by shtripat — APPROVED** [review]
Looks good. Hope CI tests would cover the path.

**Review by harshavardhana — APPROVED** [review]

**Comment by poornas** [comment]
The convoluted condition had a purpose - replication not only needs to proxy in active-active if object is missing on the site that received the request, but also another important task - it needs to figure out whether the object should be replicated [here](https://github.com/minio/minio/blob/928c0181bfaa2fbc4a4d1adc2a5106dd151de600/cmd/bucket-replication.go#L634) and [here](https://github.com/minio/minio/blob/928c0181bfaa2fbc4a4d1adc2a5106dd151de600/cmd/bucket-replication.go#L1401). 

The ProxyHeaderSet to "false" was to avoid re-proxying the request back to primary and prematurely mark replication complete by checking itself. GET/HEAD requests do not have any other identifying headers to know if it is a client request or replication request - it would be preferable to leave this as is without venturing into looking at user agent etc to demarcate a replication request

This PR needs to be reverted - otherwise, we will see the regression below again @harshavardhana 
https://github.com/minio/minio/commit/fd3f02637acf918a2db2bebede7f8bc99c1b9201

**Comment by rluetzner** [comment]
Wow. This is hard to grok. I had to write it down again for me as well.

```go
if opts.ProxyRequest || (opts.ProxyHeaderSet && !opts.ProxyRequest)
```

So either ProxyRequest is true (in which case ProxyHeaderSet is also true) or it is false, in which case we're interested if the header was present in the request.

But let's rethink this.

Header is set to "true", so we get ProxyRequest as true. The second part after `||` doesn't matter. Everything evaluates to true.

Header is missing, so ProxyRequest is false and ProxyHeaderSet is false. Everything evaluates to false.

Header is set to "false", so ProxyRequest is false, but ProxyHeaderSet evaluates to true. The whole expression evaluates to true.

@poornas  Taking all of this, I'm still convinced my change is correct or the previous code was buggy. The only difference is of the HTTP header exists or not.

**Comment by rluetzner** [comment]
In any case, even if the previous code was buggy, the fact that all places use exactly the same condition shows that it _can_ be simplified.

**Comment by poornas** [comment]
This header morphed from a 2 state boolean to track  3 states - "request was proxied", "not set", "do not proxy this" for backward compatibility and to address the regression I mentioned. 

>Header is set to "true", so we get ProxyRequest as true. The second part after || doesn't matter. Everything evaluates to true.

here  GET/HEAD is proxied from peer site

>Header is missing, so ProxyRequest is false and ProxyHeaderSet is false. Everything evaluates to false.

could be client request or replication process checking whether to replicate - if no header is set, it will again proxy back from peer site to primary based on active-active config

> Header is set to "false", so ProxyRequest is false, but ProxyHeaderSet evaluates to true. The whole expression evaluates to true.
 - this was for replication process to use and is preventing premature completion of request and proxying back to self. your PR will essentially scuttle this.

**Comment by rluetzner** [comment]
But again. If the header value is set to true, everything evaluates to true and if the header value is anything else, it'll also evaluate to true.

I can see how it should track three different States, but from the code I still don't see the difference.

**Comment by poornas** [comment]
```
if opts.ProxyRequest || (opts.ProxyHeaderSet && !opts.ProxyRequest) {
    return  //don't proxy further since this already from peer site
}
proxy if conditions applicable..
```

**Comment by rluetzner** [comment]
I've replaced this code

```
if opts.ProxyRequest || (opts.ProxyHeaderSet && !opts.ProxyRequest) {
    return  //don't proxy further since this already from peer site
}
proxy if conditions applicable..
```

with this here

```
if opts.ProxyHeaderSet {
    return  //don't proxy further since this already from peer site
}
proxy if conditions applicable..
```

I also renamed ProxyHeaderSet, which is why this might be confusing. ProxyHeaderSet is always true, when ProxyHeader is true.

And in case that ProxyHeader is false, but ProxyHeaderSet is true, the above code still does exactly the same.

You can still revert this if you want. I concede that there might be a chance that I've missed something or that you are keeping this around to be prepared for a certain situation (which I can't know).

I am however 99.9% sure that my transformation doesn't change anything.

**Pull request merged**